### DISCUSSION TITLE
What are possible module names?

### DISCUSSION THREAD

**Discussion by nnmrts** [nestdotland/nest.land#310] [Q&A] [answered]
I moved away from deno.land/x because they don't support publishing a module with hyphens in their name for no apparent reason. It seems like this is possible on nest.land, but what else is? Which characters in detail are allowed in that name string? Can we publish scoped modules with `@` and `/` like on npm?

I couldn't find any information on this. All I found was the json schema for validating the config, but that one also just says `string` at the `name` field.

Also, is there any effort to prevent name squatting and similar things?

**Comment by maximousblk** [comment, marked as answer]
Hey @nnmrts, welcome to the community!

For now, the module name can be any non-[reserved](https://github.com/nestdotland/x/blob/887e2cac2c26e2cf1a5d5b67b3af55e4cb8b7ce0/src/utils/reservedNames.ts) string with length 2 to 40 and with characters `a-z`, `A-Z`, `0-9`, `_`, `-`, `.` and `:`. characters `@` and `/` are not allowed

We're going to move to a new module naming format where all modules will be scoped by default. (`nest.land/x/[user]/[module]@[semver]/[...files]`). you'll still be able to get a top level name (`nest.land/-/[module]@[semver]/[...files]`) once your module gets manually verified and is of decent quality. this drastically eliminates top level name squatting. this is crucial because nest, by design, doesn't allow deleting the modules so we cannot follow the common practice of just deleting the squatted module.

so you can have `scope:module` formatted module names if you really want scoped names now but do note that it might cause problems migrating to new format.

you can find more info at [nestdotland/roadmap/1#notes](https://github.com/nestdotland/roadmap/issues/1/#notes)

**Reply by nnmrts** [reply]
Thanks for the answer! The scoped by default approach actually sounds awesome, exactly what I would expect from a modern module registry. I guess all current modules will be transferred to that system too, once it arrives?

What would your suggestion be if I wanted to publish an org scoped module now on nest.land to make it somewhat future-proof for any changes? I've read about the "orgs.land" in that roadmap and wonder if I can somehow prepare a module to land there once it's here. And if not, at least prepare it to be found at  `nest.land/x/[org]/[module]@[semver]/[...files]` in the future.

**Reply by maximousblk** [reply]
There's not really anything you need to do, if there will be you will be notified. if you have any modules they will automatically be moved to the new scoped URLs (without some new features which need extra steps from publishers).

we want this change to be as non breaking as possible so the old links (mostly every `GET` request to `x.nest.land/...`) will continue to work but new publishes won't be pushed there.

we don't support scoped modules currently (well `scope:module` but that's a hack) so I don't think there is a way to future proof them.

`orgs.land` is supposed to be targeted more towards "enterprise" type customers who need managed solutions like automatic audit alerts and such, but anything about orgs.land is far from confirmed. it's still just an idea and there's a chance it would never be landed so no need to be worried about that.

**Reply by nnmrts** [reply]
Alright, thanks! So I guess it's fine to register multiple accounts on nest.land, for example one for personal packages and one for org ones? Would still be nicer to just have one account and API key for multiple "usernames", but it's okay either way.
